src/dmd/dsymbolsem.d

```diff
@@ -3480,6 +3480,27 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             }
         }
 
+        if (sc.flags & (SCOPE.printf | SCOPE.scanf))
+        {
+            /* printf/scanf-like functions must be of the form:
+             *    extern (C) T printf([parameters...], const(char)* format, ...);
+             * or:
+             *    extern (C) T vprintf([parameters...], const(char)* format, va_list);
+             */
+            if (f.linkage != LINK.c ||
+                f.parameterList.length < 1 ||
+                !(f.parameterList.varargs == VarArg.variadic ||
+                  f.parameterList.varargs == VarArg.none && f.parameterList.length >= 2)
+               )
+            {
+                const p = ((sc.flags & SCOPE.printf) ? Id.printf : Id.scanf).toChars();
+                funcdecl.error("`pragma(%s)` functions must be `extern(C) T %s([parameters...], const(char)*, [..., va_list]) not `%s`",
```

Comment:
This doesn't check the last non-variadic parameter is `const(char)*`

Reply:
it does now